lib/datadog/appsec/contrib/rack/request_middleware.rb

```diff
@@ -49,6 +54,50 @@ def call(env)
             AppSec::Event.record(*both_response.map { |_action, event| event }) if both_response.any?
 
             request_return
+          ensure
+            add_waf_runtime_tags(context)
+            request!
```

Comment:
Ah, I'm very used to using bangs, so in my mind `request!` is immediately read as "mark request", which in this context is exactly equivalent to `this request has been performed`.

I understand that it may sound confusing though (probably I did too much Perl in my life and got too used to non-ambiguous inference from context).

Reply:
Also, this may tie in together with https://github.com/DataDog/dd-trace-rb/pull/1987#discussion_r864669076 as it may not be that the request has been processed that may matter but the fact that tags have been transmitted at least once.